torch/testing/_internal/common_device_type.py

```diff
@@ -335,66 +338,37 @@ def _get_tolerance_override(self, test, dtype):
+            self.precision, self.rel_tol = self._get_tolerance_override(test, dtype)
+
     # Creates device-specific tests.
     @classmethod
     def instantiate_test(cls, name, test, *, generic_cls=None):
 
-        def instantiate_test_helper(cls, name, *, test, dtype, op):
-
-            # Constructs the test's name
-            test_name = _construct_test_name(name, op, cls.device_type, dtype)
-
-            # Wraps instantiated test with op decorators
-            # NOTE: test_wrapper exists because we don't want to apply
-            #   op-specific decorators to the original test.
-            #   Test-specific decorators are applied to the original test,
-            #   however.
-            if op is not None:
-                try:
-                    active_decorators = []
-                    if op.should_skip(generic_cls.__name__, name, cls.device_type, dtype):
-                        active_decorators.append(skipIf(True, "Skipped!"))
-
-                    if op.decorators is not None:
-                        for decorator in op.decorators:
-                            # Can't use isinstance as it would cause a circular import
-                            if decorator.__class__.__name__ == 'DecorateInfo':
-                                if decorator.is_active(generic_cls.__name__, name, cls.device_type, dtype):
-                                    active_decorators += decorator.decorators
-                            else:
-                                active_decorators.append(decorator)
-
-                    @wraps(test)
-                    def test_wrapper(*args, **kwargs):
-                        return test(*args, **kwargs)
-
-                    for decorator in active_decorators:
-                        test_wrapper = decorator(test_wrapper)
-
-                    test_fn = test_wrapper
-                except Exception as ex:
-                    # Provides an error message for debugging before rethrowing the exception
-                    print("Failed to instantiate {0} for op {1}!".format(test_name, op.name))
-                    raise ex
-            else:
-                test_fn = test
-
+        def instantiate_test_helper(cls, name, *, test, param_kwargs=None):
             # Constructs the test
-            @wraps(test_fn)
-            def instantiated_test(self, name=name, test=test_fn, dtype=dtype, op=op):
-                device_arg: str = cls.get_primary_device()
-                if hasattr(test_fn, 'num_required_devices'):
-                    device_arg = cls.get_all_devices()
+            @wraps(test)
+            def instantiated_test(self, param_kwargs=param_kwargs):
+                # Add the device param kwarg if the test needs device or devices.
+                param_kwargs = {} if param_kwargs is None else param_kwargs
+                test_sig_params = inspect.signature(test).parameters
+                if 'device' in test_sig_params or 'devices' in test_sig_params:
```

Comment:
I'm not in love with the string matching but I guess it's OK as long as we document it

Reply:
True, I'm not in love with it either.. Alternatively to doing this, the `device` / `devices` arg could be unconditionally passed again. It could probably be argued that all tests instantiated with `instantiate_device_type_tests()` should accept and use `device` or `devices`, but unfortunately some tests (probably wrongly) violate this.